src/components/TodoList.tsx

```diff
@@ -0,0 +1,195 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import React, { useEffect, useRef, useState } from 'react';
+import { Todo } from '../types/Todo';
+import { USER_ID, deleteTodo, updateTodo } from '../api/todos';
+import { ErrorMessages } from '../App';
+
+type Props = {
+  todoList: Todo[];
+  todos: Todo[];
+  setTodos: (todo: Todo[]) => void;
+  setError: (error: ErrorMessages) => void;
+};
+
+export const TodoList: React.FC<Props> = ({
+  todoList = [],
+  todos,
+  setTodos,
+  setError,
+}) => {
+  // const [todos, setTodos] = useState<Todo[]>([]);
+  const [formActive, setFormActive] = useState(false);
+  const [todoTitle, setTodoTitle] = useState('');
+  const [todoId, setTodoId] = useState(Infinity);
+  const inputRef = useRef<HTMLInputElement>(null);
+  const [tempTitle, setTempTitle] = useState('');
+
+  useEffect(() => {
+    if (formActive && inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [formActive]);
+
+  function handleSetFormActive(title: string, id: number) {
+    setFormActive(true);
+    setTodoId(id);
+    setTodoTitle(title);
+    setTempTitle(title);
+  }
+
+  function handleChangeCompleted(todoSelect: Todo) {
+    setError('');
+    updateTodo({
+      id: todoSelect.id,
+      userId: USER_ID,
+      title: todoSelect.title,
+      completed: !todoSelect.completed,
+    })
+      .then(updatedTodo => {
+        setTodos(
+          [...todos].map(todo => {
+            if (todo.id === todoSelect.id) {
+              return { ...todo, completed: updatedTodo.completed };
+            }
+
+            return todo;
+          }),
+        );
+      })
```

Comment:
Use a callback to update the state `setTodos(last => [...last]...)`; this will prevent issues with asynchronous requests;
Try toggling two items as completed quickly (setting the second one before the first request is complete) and you'll see the issue: the first one will be shown as incomplete again even though it is completed in the backend